lib/spack/spack/test/cmd/env.py

```diff
@@ -1948,6 +1948,48 @@ def test_view_link_roots(tmpdir, mock_fetch, mock_packages, mock_archive,
                                  (spec.version, spec.compiler.name)))
 
 
+def test_view_link_type_copy(tmpdir, mock_fetch, mock_packages, mock_archive,
+                             install_mockery):
+    filename = str(tmpdir.join('spack.yaml'))
+    viewdir = str(tmpdir.join('view'))
+    with open(filename, 'w') as f:
+        f.write("""\
+env:
+  definitions:
+    - packages: [mpileaks, callpath]
+    - compilers: ['%%gcc', '%%clang']
+  specs:
+    - matrix:
+        - [$packages]
+        - [$compilers]
+
+  view:
+    combinatorial:
+      root: %s
+      select: ['%%gcc']
+      exclude: [callpath]
+      link_type: copy
+      link: 'roots'
+      projections:
+        'all': '{name}/{version}-{compiler.name}'""" % viewdir)
+    with tmpdir.as_cwd():
+        env('create', 'test', './spack.yaml')
+        with ev.read('test'):
+            install()
+
+        test = ev.read('test')
+        for spec in test._get_environment_specs():
+            if spec in test.roots() and (spec.satisfies('%gcc') and
+                                         not spec.satisfies('callpath')):
+                assert os.path.exists(
+                    os.path.join(viewdir, spec.name, '%s-%s' %
+                                 (spec.version, spec.compiler.name)))
+            else:
+                assert not os.path.exists(
+                    os.path.join(viewdir, spec.name, '%s-%s' %
+                                 (spec.version, spec.compiler.name)))
```

Comment:
Shouldn't this test assert that the files are actually copied in addition to the fact that they exist?

I would verify `not os.path.islink()` for these -- to ensure that it's not just doing normal symlinking.

Reply:
Yeah the link test is a good idea. In terms of testing that files are actually copied, do the mocks actually generate files to check inside the install directories?  I started diving into that but didn't get very far.

Reply:
The tricky thing here is that the first file location you'd test is a link. For example in my test environment above the top directories `copy`, `hardlink` and `symlink` are all links. The distinction is the files inside `copy/bin` aren't links and the files inside `symlink/bin` are.  I've tried adding:
```
        test = ev.read('test')
        for spec in test._get_environment_specs():
            if spec in test.roots() and (spec.satisfies('%gcc') and
                                         not spec.satisfies('callpath')):
                assert os.path.exists(
                    os.path.join(viewdir, spec.name, '%s-%s' %
                                 (spec.version, spec.compiler.name), 'bin'))
                assert not os.path.islink(
                    os.path.join(viewdir, spec.name, '%s-%s' %
                                 (spec.version, spec.compiler.name), 'bin')
```
but it looks like the mock's aren't creating the `bin` directory.

Reply:
Pretty sure I got this figured out with @becker33's help on slack.  I added two more unit tests to check all three link types.